Blueprints: Add handler for creating new blueprint

amirfefer opened the pull request:
This PR adds a handler for submitting  the blueprint's wizard, by implementing a mapper from redux store (wizard slice) to a valid blueprint creation request.

https://github.com/RedHatInsights/image-builder-frontend/assets/11807069/4264d958-46c7-49c4-a548-383ea900390b

amirfefer commented:
I created a custom `Footer` for the review step, following the mocks:

<img width="518" alt="Screenshot 2024-02-01 at 17 52 44" src="https://github.com/RedHatInsights/image-builder-frontend/assets/11807069/e3bb5687-06ee-4d6e-a75b-94e11bd59a55">

ezr-ondrej commented on src/store/enhancedImageBuilderApi.ts:
```diff
@@ -62,7 +87,7 @@ const enhancedApi = imageBuilderApi.enhanceEndpoints({
             // Typescript is unaware of tag types being defined concurrently in enhanceEndpoints()
             // @ts-expect-error
-            dispatch(imageBuilderApi.util.invalidateTags(['Compose']));
+            dispatch(imageBuilderApi.util.invalidateTags(['Blueprints']));
```
not sure, if we want this, but if we do, should we invalidate both of the tags?

amirfefer replied:
great rebase issues catch! :D

ezr-ondrej commented on src/Components/CreateImageWizardV2/steps/Review/Footer.tsx:
```diff
@@ -0,0 +1,92 @@
+  isLoading?: boolean;
+};
+
+const FinishedWizardFooter = ({
```
nitpick
```suggestion
const ReviewStepFooter = ({
```

ezr-ondrej reviewed (commented):
Nice, great work here :)

Few founds:
* ~It doesn't save the blueprint for me, with following error:~ EDIT: This is caused by the backend, so this is out of scope of this PR
   ```
    {
      "errors": [
          {
              "detail": "request body has an error: doesn't match schema #/components/schemas/CreateBlueprintRequest: Error at \"/image_requests/1/upload_request/type\": value \"image-installer\" is not one of the allowed values\nSchema:\n  {\n    \"enum\": [\n      \"aws\",\n      \"gcp\",\n      \"azure\",\n      \"aws.s3\",\n      \"oci.objectstorage\"\n    ],\n    \"type\": \"string\"\n  }\n\nValue:\n  \"image-installer\"\n",
              "title": "400"
          }
      ]
  }
  ```
* The image name was required to be lower-cased without spaces, we now impose it on Blueprints, but that's probably not what we want to do, it's fine for now tho, lets have that discussion outside of scope of this PR
* General thought to discuss with UX: Can we make the save button clickable by default (to save the blueprint and require double click only for save and build images)?

Few nitpicks in the code.

amirfefer commented:
Thanks @ezr-ondrej 👍 
> It doesn't save the blueprint for me, with following error:

What types of images you tried to add in the blueprint request?

ezr-ondrej commented:
> What types of images you tried to add in the blueprint request?

AWS and baremetal. Baremetal might have been the offender

amirfefer commented:
Thanks @ezr-ondrej! I addressed your comments :)
The wizard is still in progress feature, so some gaps would happen, the handler doesn't handle all cases.
But I found an issue with wrong type casting, which might caused your error

ezr-ondrej commented on src/store/enhancedImageBuilderApi.ts:
```diff
@@ -20,6 +20,31 @@ const enhancedApi = imageBuilderApi.enhanceEndpoints({
+      onQueryStarted: async (_, { dispatch, queryFulfilled }) => {
+        queryFulfilled
+          .then(() => {
+            // Typescript is unaware of tag types being defined concurrently in enhanceEndpoints()
+            // @ts-expect-error
```
linter in the iqe pipeline is saying this is unused (meaning ts would not error?)

amirfefer commented on src/test/Components/CreateImageWizardV2/CreateImageWizard.test.tsx:
```diff
@@ -952,6 +952,10 @@ describe('Step Upload to AWS', () => {
       // skip repositories
       await clickNext();
       // skip Details
+      const blueprintName = await screen.findByRole('textbox', {
+        name: /blueprint name/i,
+      });
+      await user.type(blueprintName, 'valid-name');
```
this block is added because blueprint name is required, a bit long test file :sweat_smile:  later can be refactored to be more dry.

amirfefer commented:
/retest

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,150 @@
+  const customizations = getCustomizations(wizardState, orgID);
+
+  return {
+    name: wizardState.details.blueprintName,
```
For the name, the description, and distribution can we use a selector function instead? We will just need to define custom selectors in the wizardSlice for this. That way the shape of the wizardState is decoupled from the rest of the application. So something like:
```javascript
export const selectBlueprintName = (state: RootState) => {
  return state.details.blueprintName;
};
```

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,150 @@
+  };
+};
+
+const getImageRequest = (state: wizardState): ImageRequest[] => {
```
Since this is an array can it be `getImageRequests` (plural)?

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,150 @@
+};
+const getImageOptions = (
+  imageType: string,
+  state: wizardState
```
Let's use selector functions for everything we access from the state here. There are already existing selector functions for a lot of these (if not all).

amirfefer commented:
@lucasgarfield, Thanks for your review!

Selectors should be used only  in the scope of  react components, for re-rendering optimizations.  I extracted the state-to-request mapper to be outside of a component scope, due to coupling reasons and re-use,  we don't  need re-rendering optimizations in submission handlers, it just add tones of boilerplate with out any benefits.

amirfefer commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,150 @@
+      return 'aws.s3';
+    case 'ami':
+      return 'aws';
+    default: {
```
This might  happen. because edge images haven't implemented yet, returning a string is forced by the image builder API, I'd prefer to return undefined and raise an error.  The wizard is in progress work under experimental flag, there's a TODO for the missing `edge` types, I think throwing an error would be better than just a warning 👍

lucasgarfield commented on src/test/Components/CreateImageWizardV2/CreateImageWizard.test.tsx:
```diff
@@ -884,7 +884,7 @@ describe('Step Upload to AWS', () => {
         name: /blueprint name/i,
       });
       // 64 character name
```
Can we just delete this silly comment? :sweat_smile:

amirfefer commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,150 @@
+    files: undefined,
+    subscription: getSubscription(state, orgID),
+    packages: undefined,
+    payload_repositories: undefined,
```
I guess it hasn't implemented yet? the wizardState has only
```js
 repositories: {
    customRepositories: CustomRepository[];
  };
```

lucasgarfield replied:
Hm, you're right. Let's wait for now, sorry for the false alarm. :sweat_smile:

lucasgarfield commented on src/Components/CreateImageWizardV2/CreateImageWizard.tsx:
```diff
@@ -44,21 +48,37 @@ import { ImageBuilderHeader } from '../sharedComponents/ImageBuilderHeader';
   disableNext: disableNext,
+  nextButtonText = 'Next',
+  isLoading,
+  onSubmit,
```
I might be missing something, but are any of these changes to `<CustomWizardFooter>` used in this pull request? It kind of looks like it might be from an early iteration before you extracted everything into `<ReviewWizardFooter>` but maybe I'm not seeing something...

lucasgarfield replied:
And FWIW I think having a separate component `<ReviewWizardFooter>` is exactly what we should do, these footers are quite small and simple so it is better to just have separate components.

amirfefer replied:
Yes, it was a leftover from early iteration, I kept the ouid tho.

lucasgarfield commented on src/Components/CreateImageWizardV2/CreateImageWizard.tsx:
```diff
@@ -212,7 +252,12 @@ const CreateImageWizard = () => {
+            footer={
+              <ReviewWizardFooter
+                isLoading={isCreationLoading}
+                onSubmit={onSubmit}
```
We could get rid of these two props if we just moved the `useMutation` hook into `<ReviewWizardFooter>` and also moved the `onSubmit` handler there as well. 

Generally I've tried to avoid passing down things like `isLoading` as props... since they are cached by RTKQ, it is cheap to just get them by calling the hook in the component where they are being used. The goal being to avoid having to search through the code to figure out exactly _what_ is loading. In this case if you are in the `<ReviewWizardFooter>` component you see that an `isLoading` prop gets passed in, but to figure out exactly _what_ is loading you have to go back to this component (which is already getting quite large and messy). If we just call the `useMutation` hook inside `<ReviewWizardFooter>` then we can more quickly understand where `isLoading` is coming from. Hope that makes sense...

amirfefer replied:
Moved to `<ReviewWizardFooter>`

lucasgarfield commented on src/Components/CreateImageWizardV2/steps/Review/Footer.tsx:
```diff
@@ -0,0 +1,92 @@
+  const { goToPrevStep, close } = useWizardContext();
+  const [isOpen, setIsOpen] = useState(false);
+  const SAVE = 'save';
+  const BUILD = 'build';
```
We could gain a bit of additional type safety by using a type here instead of constants I think:
```typescript
type submitType = 'save' | 'build'
```

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,164 @@
+  switch (imageType) {
+    case 'aws':
+      if (selectAwsShareMethod(state) === 'sources')
+        return { share_with_sources: [selectAwsSource(state)?.id || ''] };
```
I'm not sure about the `|| ''` here... in the state, `aws.source` can be either a `V1ListSourceResponseItem` or `undefined`. And the `id` field is optional on that type. So we do need to handle the case where there is not an id.

But I'm not sure passing an empty string is what we should do. After we build an AWS image, we upload it to an S3 bucket and share it with that AWS account via the id. If there is no id, then we won't be able to share the image, and it will be essentially inaccessible. 

So I think we probably ought to fail in that case. 

I need to think this through a bit more though.

@croissanne can I get a sanity check?

croissanne replied:
yea for AWS if there's nothing to share it with, it's not accessible.

lucasgarfield replied:
Thanks @croissanne!

We will need to ensure that we check that there is an ID on the AWS step (I created an issue for that #1637).  

There are other edge cases to consider related to this but ultimately, we should hopefully never get to this point... if we do, it's a bug. Maybe it would be easier to track down the source of the bug if we just throw an error here, something like:
```typescript
throw (
  new Error('AWS account ID could not be resolved from source')
);
```

What do you think @amirfefer ?

amirfefer replied:
How about using `required` validation in the AWS step for the account number (if the sources radio is chosen)

<img width="982" alt="Screenshot 2024-02-08 at 20 18 02" src="https://github.com/osbuild/image-builder-frontend/assets/11807069/3da1db4d-aef4-4d57-b0af-836d0c4c279c">

I guess a  backend validation is needed as well?

lucasgarfield replied:
Yes, we'll definitely add that to the AWS step!

lucasgarfield replied:
But we still need to make Typescript happy and I don't think using an empty string is the right way... does throwing an error sound good to you? Like I mentioned below, I'd rather fail catastrophically and immediately rather than build an image that the user won't be able to access later...

amirfefer commented:
Thanks @lucasgarfield!
I added also the `Save and build` implementation to the handler :)

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,164 @@
+
+const getSubscription = (state: RootState, orgID: string): Subscription => {
+  const initialSubscription = {
+    'activation-key': selectActivationKey(state) || '',
```
Similar to the discussion around the AWS source ID, I don't think we want to use an empty string here because if we get to this point and there is no activation key... it is a bug. We should have validation on the activation keys step that prevents that from ever happening, so if it does, I think I would prefer that we just fail catastrophically rather than failing 'silently' by building an image that won't have a working subscription.

lucasgarfield commented on src/Components/CreateImageWizardV2/utilities/requestMapper.tsx:
```diff
@@ -0,0 +1,164 @@
+    containers: undefined,
+    directories: undefined,
+    files: undefined,
+    subscription: getSubscription(state, orgID),
```
We'll also need to handle the case where a user selects "Register later"... I think we would just keep the subscription as `undefined` in that case.

lucasgarfield commented:
I think this pull request is coming along nicely. I keep coming back with more and more comments but we support so many possible customizations and states, it is just really really tricky to put the request together. Thanks for bearing with me. :sweat_smile:

lucasgarfield commented on src/Components/CreateImageWizardV2/steps/Review/Footer.tsx:
```diff
@@ -0,0 +1,115 @@
+      requestBody &&
+      (await createBlueprint({ createBlueprintRequest: requestBody }).unwrap());
+    blueprint && buildBlueprint({ id: blueprint.id });
+    setIsOpen(false);
```
This is also being set to false a few lines above, one of them is probably a duplicate.

lucasgarfield commented:
/retest

ezr-ondrej commented:
/retest
it was supposed to be fixed! :'(

lucasgarfield reviewed (approved):
I'm going to merge this, great work @amirfefer. There are a few things I want to deal with in follow-up PRs but I don't think there's a good reason to delay this.
